### ISSUE TITLE
Question regarding "class agnostic center heatmap" P(t-tao)_agn

### ISSUE THREAD

**Issue by NosremeC** [JialianW/TraDeS#8]
Hi there, thank you for sharing your great work! I have a question here, please do correct me if I got it wrong:

We use a "class agnostic center heatmap"(of previous feat) P_agn times the previous feat, then passing it to the DCN along with the offset OD to get the "center attentive feature", which can be used to enhance the current feat. In your paper, it said P_agn was obtained from the heatmap output of CenterNet, but in the base_model section in your training code I found that the P_agn, aka 'pre_hm_i', was always generated directly from the ground truth.

base_model.py:
![1](https://user-images.githubusercontent.com/68608242/112924417-11297080-90de-11eb-952c-41c12fdd6fd8.jpg)
trainer.py:
![2](https://user-images.githubusercontent.com/68608242/112926157-09b79680-90e1-11eb-9bc7-bfea1501f050.jpg)
generic_dataset.py, __getitem__:
![3](https://user-images.githubusercontent.com/68608242/112926191-15a35880-90e1-11eb-889e-8f74259ddc3c.jpg)
_get_pre_dets:
![5](https://user-images.githubusercontent.com/68608242/112926340-4daa9b80-90e1-11eb-8b90-6d4ecc53a747.jpg)

I also read your code for test, the Detector seemed ok since it used the previous heatmap result. Have I misunderstood anything here? Many thanks!

**Comment by JialianW** [comment]
Thanks for your interest in our work!

We use the ground truth heatmap in training, while in testing we use the heatmap predicted by the model.

**Comment by NosremeC** [comment]
> Thanks for your interest in our work!
> 
> We use the ground truth heatmap in training, while in testing we use the heatmap predicted by the model.

So you mean you have used the gt heatmap to make the feature center-attentive during training?  Wouldn't that affect the performance of testing? Since during testing there is no gt so we have to rely on the previous centerNet output. Shouldn't you also train the model in the same way in which it uses P_agn from the previous result and as the loss goes down the performance of centerNet will improve?

**Comment by JialianW** [comment]
Good question. I cited the explanation from the CenterTrack paper for your reference:

![1617075731(1)](https://user-images.githubusercontent.com/39213200/112930652-b3e6ec80-90e8-11eb-8413-1661fa96e5da.png)

**Comment by NosremeC** [comment]
> Good question. I cited the explanation from the CenterTrack paper for your reference:
> 
> ![1617075731(1)](https://user-images.githubusercontent.com/39213200/112930652-b3e6ec80-90e8-11eb-8413-1661fa96e5da.png)

That is very interesting. Now I see what is actually happening during the training, but still I don't fully understand why we don't just use the real heatmap output to train instead of manually introducing errors. I also read through the ablation study of centertrack and they never explained why abandoning using the real heatmap, maybe it creates way too much noise during the early stage of training when the detection head is still weak? When the detection head outputs a bad heatmap it will degrade the quality of the original feature after applying it, and the result will further make the detection on the current feature even worse after they are interacted, so the training somehow gets stuck in the early stage? 

Thank you for sharing the reference though!

**Comment by JialianW** [comment]
What you had said could be a reason. Not sure using the real heatmap or manipulated gt would make difference in performance. The previous heatmap does not play that important role in the method. The manipulated gt should be ok enough.

**Comment by NosremeC** [comment]
> What you had said could be a reason. Not sure using the real heatmap or manipulated gt would make difference in performance. The previous heatmap does not play that important role in the method. The manipulated gt should be ok enough.

Totally make sense! I'll just move on from here now. Thank you for the patience!
